common/lib/client/rest.ts

```diff
@@ -0,0 +1,225 @@
+				throw new ErrorInfo('clientId must be either a string or null', 40012, 400);
+			else if(normalOptions.clientId === '*')
+				throw new ErrorInfo('Can’t use "*" as a clientId as that string is reserved. (To change the default token request behaviour to use a wildcard clientId, use {defaultTokenParams: {clientId: "*"}})', 40012, 400);
+		}
+
+		Logger.logAction(Logger.LOG_MINOR, 'Rest()', 'started; version = ' + Defaults.version);
+
+		this.baseUri = this.authority = function(host) { return Defaults.getHttpScheme(normalOptions) + host + ':' + Defaults.getPort(normalOptions, false); };
+		this._currentFallback = null;
+
+		this.serverTimeOffset = null;
+		this.auth = new Auth(this, normalOptions);
+		this.channels = new Channels(this);
+		this.push = new Push(this);
+	}
+
+	stats(params: RequestParams, callback: StandardCallback<PaginatedResult<Stats>>): Promise<PaginatedResult<Stats>> | void {
+		/* params and callback are optional; see if params contains the callback */
+		if(callback === undefined) {
+			if(typeof(params) == 'function') {
+				callback = params;
+				params = null;
+			} else {
+				if(this.options.promises) {
+					return Utils.promisify(this, 'stats', arguments) as Promise<PaginatedResult<Stats>>;
+				}
+				callback = noop;
+			}
+		}
+		const headers = Utils.defaultGetHeaders(),
+			format = this.options.useBinaryProtocol ? Utils.Format.msgpack : Utils.Format.json,
+			envelope = Http.supportsLinkHeaders ? undefined : format;
+
+		if(this.options.headers)
+			Utils.mixin(headers, this.options.headers);
+
+		(new PaginatedResource(this, '/stats', headers, envelope, function(body: unknown, headers: Record<string, string>, unpacked?: boolean) {
+			const statsValues = (unpacked ? body : JSON.parse(body as string));
+			for(let i = 0; i < statsValues.length; i++) statsValues[i] = Stats.fromValues(statsValues[i]);
+			return statsValues;
+		})).get(params as Record<string, string>, callback);
+	}
+
+	time(params: RequestParams | StandardCallback<number>, callback?: StandardCallback<number>): Promise<number> | void {
+		/* params and callback are optional; see if params contains the callback */
+		if(callback === undefined) {
+			if(typeof(params) == 'function') {
+				callback = params;
+				params = null;
+			} else {
+				if(this.options.promises) {
+					return Utils.promisify(this, 'time', arguments) as Promise<number>;
+				}
+			}
+		}
+
+		const _callback = callback || noop;
+
+		const headers = Utils.defaultGetHeaders();
+		if(this.options.headers)
+			Utils.mixin(headers, this.options.headers);
+		const timeUri = (host: string) => { return this.authority(host) + '/time' };
+		Http.get(this, timeUri, headers, params as RequestParams, (err?: ErrorInfo | ErrnoException | null, res?: unknown, headers?: Record<string, string>, unpacked?: boolean) => {
+			if(err) {
+				_callback(err);
+				return;
+			}
+			if(!unpacked) res = JSON.parse(res as string);
+			const time = (res as number[])[0];
```

Comment:
Unsafe access to array (and potentially unsafe cast?) ☝️

Reply:
the REST API will always return an array with a single number here so I think it's safe enough

Reply:
Resolving this conversation on @ikbalkaya's behalf.

(though I would note that I do have sympathy with the opening comment... an API boundary is a place where we _should_ be more diligent in checking things / not making assumptions of correctness of external systems)